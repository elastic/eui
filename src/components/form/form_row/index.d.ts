import { CommonProps, ExclusiveUnion } from '../../common';

import {
  FunctionComponent,
  ReactNode,
  ReactElement,
  HTMLAttributes,
} from 'react';

declare module '@elastic/eui' {
  /**
   * @see './form_row.js'
   */
  export type EuiFormRowCommonProps = CommonProps & {
    error?: ReactNode | ReactNode[];
    fullWidth?: boolean;
    hasEmptyLabelSpace?: boolean;
    helpText?: ReactNode;
    isInvalid?: boolean;
    label?: ReactNode;
    labelAppend?: ReactNode;
    describedByIds?: string[];
    compressed?: boolean;
    display?: 'row' | 'rowCompressed' | 'columnCompressed';
    displayOnly?: boolean;
  };

  type LabelProps = {
    labelType?: 'label';
  } & EuiFormRowCommonProps &
    HTMLAttributes<HTMLDivElement>;

  type LegendProps = {
    labelType?: 'legend';
  } & EuiFormRowCommonProps &
    HTMLAttributes<HTMLFieldSetElement>;

  export type EuiFormRowProps = ExclusiveUnion<LabelProps, LegendProps> & {
    children: ReactElement;
  };

  export const EuiFormRow: FunctionComponent<EuiFormRowProps>;
}
