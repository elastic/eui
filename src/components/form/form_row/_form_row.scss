/**
 * 1. Coerce inline form elements to behave as block-level elements.
 * 2. For inline forms, we need to add margin if the label doesn't exist.
 */
.euiFormRow {
  display: flex; /* 1 */
  flex-direction: column; /* 1 */
  max-width: $euiFormMaxWidth;

  + .euiFormRow {
    margin-top: $euiSize;
  }
}

.euiFormRow--fullWidth {
  max-width: 100%;
}

.euiFormRow--hasEmptyLabelSpace {
  margin-top: ($euiFontSizeXS * $euiLineHeight) + $euiSizeXS; /* 2 */
  // the following ensure that contents that aren't inheritly the same height
  // as inputs will align to the vertical center
  min-height: $euiFormControlHeight;
  padding-bottom: 0;
  justify-content: center;
}

.euiFormRow__labelWrapper {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  margin-bottom: $euiSizeXS;
}

.euiFormRow--horizontal {
  flex-direction: row;
  align-items: stretch;

  .euiFormRow__label {
    @include euiTextBreakWord;
    hyphens: auto;
    max-width: 100%; // Fixes IE
  }

  .euiFormRow__labelWrapper {
    display: block;
    line-height: $euiFormControlCompressedHeight - 1px; // The 1px less helps the alignment of the text baseline
    width: calc(33% - #{$euiSizeS});
    margin-right: $euiSizeS;
    margin-bottom: 0;
  }

  .euiFormRow__fieldWrapper {
    width: 67%;
  }

  + .euiFormRow--horizontal {
    margin-top: $euiSizeS;
  }
}

.euiFormRow__fieldWrapperDisplayOnly {
  min-height: $euiFormControlHeight;
  display: flex;
  align-items: center;
}

.euiFormRow--compressed {
  &.euiFormRow--hasEmptyLabelSpace {
    min-height: $euiFormControlCompressedHeight;
  }

  .euiFormRow__fieldWrapperDisplayOnly {
    min-height: $euiFormControlCompressedHeight;
  }
}
