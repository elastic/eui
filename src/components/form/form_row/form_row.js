import React, {
  cloneElement,
  Component,
} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { get } from 'lodash';

import { EuiFormHelpText } from '../form_help_text';
import { EuiFormErrorText } from '../form_error_text';
import { EuiFormLabel } from '../form_label';

import makeId from './make_id';

export class EuiFormRow extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isFocused: false,
      id: props.id || makeId()
    };

    this.onFocus = this.onFocus.bind(this);
    this.onBlur = this.onBlur.bind(this);
  }

  onFocus(...args) {
    // Doing this to allow onFocus to be called correctly from the child input element as this component overrides it
    const onChildFocus = get(this.props, 'children.props.onFocus');
    if (onChildFocus) {
      onChildFocus(...args);

    }

    this.setState({
      isFocused: true,
    });
  }

  onBlur(...args) {
    // Doing this to allow onBlur to be called correctly from the child input element as this component overrides it
    const onChildBlur = get(this.props, 'children.props.onBlur');
    if (onChildBlur) {
      onChildBlur(...args);

    }

    this.setState({
      isFocused: false,
    });
  }
  render() {
    const {
      children,
      helpText,
      isInvalid,
      error,
      label,
      hasEmptyLabelSpace,
      fullWidth,
      className,
      ...rest
    } = this.props;

    const { id } = this.state;

    const classes = classNames(
      'euiFormRow',
      {
        'euiFormRow--hasEmptyLabelSpace': hasEmptyLabelSpace,
        'euiFormRow--fullWidth': fullWidth,
      },
      className
    );

    let optionalHelpText;

    if (helpText) {
      optionalHelpText = (
        <EuiFormHelpText id={`${id}-help`} className="euiFormRow__text">
          {helpText}
        </EuiFormHelpText>
      );
    }

    let optionalErrors;

    if (error && isInvalid) {
      const errorTexts = Array.isArray(error) ? error : [error];
      optionalErrors = errorTexts.map((error, i) => (
        <EuiFormErrorText key={error} id={`${id}-error-${i}`} className="euiFormRow__text">
          {error}
        </EuiFormErrorText>
      ));
    }

    let optionalLabel;

    if (label) {
      optionalLabel = (
        <EuiFormLabel
          isFocused={this.state.isFocused}
          isInvalid={isInvalid}
          htmlFor={id}
        >
          {label}
        </EuiFormLabel>
      );
    }

    const describingIds = [];
    if (optionalHelpText) {
      describingIds.push(optionalHelpText.props.id);
    }
    if (optionalErrors) {
      optionalErrors.forEach(error => describingIds.push(error.props.id));
    }

    const optionalProps = {};
    if (describingIds.length > 0) {
      optionalProps[`aria-describedby`] = describingIds.join(` `);
    }

    const field = cloneElement(children, {
      id,
      onFocus: this.onFocus,
      onBlur: this.onBlur,
      ...optionalProps
    });

    return (
      <div
        className={classes}
        {...rest}
        id={`${id}-row`}
      >
        {optionalLabel}
        {field}
        {optionalErrors}
        {optionalHelpText}
      </div>
    );
  }
}

EuiFormRow.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  label: PropTypes.node,
  id: PropTypes.string,
  isInvalid: PropTypes.bool,
  error: PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.string)]),
  helpText: PropTypes.node,
  hasEmptyLabelSpace: PropTypes.bool,
  fullWidth: PropTypes.bool,
};

EuiFormRow.defaultProps = {
  hasEmptyLabelSpace: false,
  fullWidth: false,
};
