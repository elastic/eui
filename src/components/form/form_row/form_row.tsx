/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React, {
  cloneElement,
  Component,
  Children,
  HTMLAttributes,
  ReactElement,
  ReactNode,
} from 'react';
import classNames from 'classnames';
import { ExclusiveUnion, CommonProps, keysOf } from '../../common';

import { get } from '../../../services/objects';

import { EuiFormHelpText } from '../form_help_text';
import { EuiFormErrorText } from '../form_error_text';
import { EuiFormLabel } from '../form_label';

import { htmlIdGenerator } from '../../../services/accessibility';

const displayToClassNameMap = {
  row: null,
  rowCompressed: 'euiFormRow--compressed',
  columnCompressed: 'euiFormRow--compressed euiFormRow--horizontal',
  center: null,
  centerCompressed: 'euiFormRow--compressed',
  columnCompressedSwitch:
    'euiFormRow--compressed euiFormRow--horizontal euiFormRow--hasSwitch',
};

export const DISPLAYS = keysOf(displayToClassNameMap);

export type EuiFormRowDisplayKeys = keyof typeof displayToClassNameMap;

interface EuiFormRowState {
  isFocused: boolean;
  id: string;
}

type EuiFormRowCommonProps = CommonProps & {
  /**
   * When `rowCompressed`, just tightens up the spacing;
   * Set to `columnCompressed` if compressed
   * and horizontal layout is needed.
   * Set to `center` or `centerCompressed` to align non-input
   * content better with inline rows.
   * Set to `columnCompressedSwitch` if the form control being passed
   * as the child is a switch.
   */
  display?: EuiFormRowDisplayKeys;
  hasEmptyLabelSpace?: boolean;
  fullWidth?: boolean;
  /**
   * IDs of additional elements that should be part of children's `aria-describedby`
   */
  describedByIds?: string[];
  /**
   * Escape hatch to not render duplicate labels if the child also renders a label
   */
  hasChildLabel?: boolean;
  /**
   * ReactElement to render as this component's content
   */
  children: ReactElement;
  label?: ReactNode;
  /**
   * Adds an extra node to the right of the form label without
   * being contained inside the form label. Good for things
   * like documentation links.
   */
  labelAppend?: any;
  id?: string;
  isInvalid?: boolean;
  error?: ReactNode | ReactNode[];
  /**
   *  Adds a single node/string or an array of nodes/strings below the input
   */
  helpText?: ReactNode | ReactNode[];
};

type LabelProps = {
  labelType?: 'label';
} & EuiFormRowCommonProps &
  HTMLAttributes<HTMLDivElement>;

type LegendProps = {
  /**
   * Defaults to rendering a `<label>` but if passed `'legend'` for labelType,
   * will render both a `<legend>` and the surrounding container as a `<fieldset>`
   */
  labelType?: 'legend';
} & EuiFormRowCommonProps &
  HTMLAttributes<HTMLFieldSetElement>;

export type EuiFormRowProps = ExclusiveUnion<LabelProps, LegendProps>;

export class EuiFormRow extends Component<EuiFormRowProps, EuiFormRowState> {
  static defaultProps = {
    display: 'row',
    hasEmptyLabelSpace: false,
    fullWidth: false,
    describedByIds: [],
    labelType: 'label',
    hasChildLabel: true,
  };

  state: EuiFormRowState = {
    isFocused: false,
    id: this.props.id || htmlIdGenerator()(),
  };

  onFocus = (...args: any[]) => {
    // Doing this to allow onFocus to be called correctly from the child input element as this component overrides it
    const onChildFocus = get(this.props, 'children.props.onFocus');
    if (onChildFocus) {
      onChildFocus(...args);
    }

    this.setState(({ isFocused }) => {
      if (!isFocused) {
        return {
          isFocused: true,
        };
      } else {
        return null;
      }
    });
  };

  onBlur = (...args: any[]) => {
    // Doing this to allow onBlur to be called correctly from the child input element as this component overrides it
    const onChildBlur = get(this.props, 'children.props.onBlur');
    if (onChildBlur) {
      onChildBlur(...args);
    }

    this.setState({
      isFocused: false,
    });
  };

  render() {
    const {
      children,
      helpText,
      isInvalid,
      error,
      label,
      labelType,
      labelAppend,
      hasEmptyLabelSpace,
      fullWidth,
      className,
      describedByIds,
      display,
      hasChildLabel,
      id: propsId,
      ...rest
    } = this.props;

    const { id } = this.state;

    const classes = classNames(
      'euiFormRow',
      {
        'euiFormRow--hasEmptyLabelSpace': hasEmptyLabelSpace,
        'euiFormRow--fullWidth': fullWidth,
      },
      displayToClassNameMap[display!], // Safe use of ! as default prop is 'row'
      className
    );

    let optionalHelpTexts;

    if (helpText) {
      const helpTexts = Array.isArray(helpText) ? helpText : [helpText];
      optionalHelpTexts = helpTexts.map((helpText, i) => {
        const key = typeof helpText === 'string' ? helpText : i;
        return (
          <EuiFormHelpText
            key={key}
            id={`${id}-help-${i}`}
            className="euiFormRow__text">
            {helpText}
          </EuiFormHelpText>
        );
      });
    }

    let optionalErrors;

    if (error && isInvalid) {
      const errorTexts = Array.isArray(error) ? error : [error];
      optionalErrors = errorTexts.map((error, i) => {
        const key = typeof error === 'string' ? error : i;
        return (
          <EuiFormErrorText
            key={key}
            id={`${id}-error-${i}`}
            className="euiFormRow__text">
            {error}
          </EuiFormErrorText>
        );
      });
    }

    let optionalLabel;
    const isLegend = label && labelType === 'legend' ? true : false;

    if (label || labelAppend) {
      let labelProps = {};
      if (isLegend) {
        labelProps = {
          type: labelType,
        };
      } else {
        labelProps = {
          htmlFor: hasChildLabel ? id : undefined,
          isFocused: this.state.isFocused,
          type: labelType,
        };
      }
      optionalLabel = (
        <div className="euiFormRow__labelWrapper">
          <EuiFormLabel
            className="euiFormRow__label"
            isInvalid={isInvalid}
            aria-invalid={isInvalid}
            {...labelProps}>
            {label}
          </EuiFormLabel>
          {labelAppend && ' '}
          {labelAppend}
        </div>
      );
    }

    const optionalProps: React.AriaAttributes = {};
    /**
     * Safe use of ! as default prop is []
     */
    const describingIds = [...describedByIds!];

    if (optionalHelpTexts) {
      optionalHelpTexts.forEach((optionalHelpText) =>
        describingIds.push(optionalHelpText.props.id)
      );
    }

    if (optionalErrors) {
      optionalErrors.forEach((error) => describingIds.push(error.props.id));
    }

    if (describingIds.length > 0) {
      optionalProps['aria-describedby'] = describingIds.join(' ');
    }

    const field = cloneElement(Children.only(children), {
      id,
      onFocus: this.onFocus,
      onBlur: this.onBlur,
      ...optionalProps,
    });

    const fieldWrapperClasses = classNames('euiFormRow__fieldWrapper', {
      euiFormRow__fieldWrapperDisplayOnly:
        /**
         * Safe use of ! as default prop is 'row'
         */
        display!.startsWith('center'),
    });

    const sharedProps = {
      className: classes,
      id: `${id}-row`,
    };

    const contents = (
      <React.Fragment>
        {optionalLabel}
        <div className={fieldWrapperClasses}>
          {field}
          {optionalErrors}
          {optionalHelpTexts}
        </div>
      </React.Fragment>
    );

    return labelType === 'legend' ? (
      <fieldset
        {...sharedProps}
        {...(rest as HTMLAttributes<HTMLFieldSetElement>)}>
        {contents}
      </fieldset>
    ) : (
      <div {...sharedProps} {...(rest as HTMLAttributes<HTMLDivElement>)}>
        {contents}
      </div>
    );
  }
}
