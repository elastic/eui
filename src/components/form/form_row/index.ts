export { EuiFormRow } from './form_row';
