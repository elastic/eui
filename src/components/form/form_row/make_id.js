export default function makeId() {
  return Math.random().toString(36).slice(-8);
}
