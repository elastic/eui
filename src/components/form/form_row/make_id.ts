/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

// Generate statistically almost-certainly-unique `id`s for associating form
// inputs with their labels and other descriptive text elements.
function makeId(): string {
  console.log(
    'WARNING: makeId is deprecated. Use htmlIdGenerator from @elastic/eui instead.'
  );
  return Math.random().toString(36).slice(-8);
}

export default makeId;
