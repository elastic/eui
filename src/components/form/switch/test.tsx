import React from 'react';

import { EuiSwitch } from './switch';

export const Test = ({ checked }: { checked: boolean }) => (
  <EuiSwitch
    checked={checked}
    label="Test"
    onChange={e => console.log(e.target.checked)}
  />
);
