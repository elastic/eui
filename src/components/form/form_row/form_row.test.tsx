/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { shallow, render, mount } from 'enzyme';
import { requiredProps } from '../../../test';

import { EuiFormRow, DISPLAYS } from './form_row';

describe('EuiFormRow', () => {
  test('is rendered', () => {
    const component = render(
      <EuiFormRow {...requiredProps}>
        <input />
      </EuiFormRow>
    );

    expect(component).toMatchSnapshot();
  });

  test('ties together parts for accessibility', () => {
    const props = {
      label: 'Label',
      helpText: 'Help text',
      isInvalid: true,
      error: ['Error one', 'Error two'],
    };

    const tree = shallow(
      <EuiFormRow {...requiredProps} {...props}>
        <input />
      </EuiFormRow>
    );

    // Input is labeled by the label.
    expect(tree.find('input').prop('id')).toEqual('generated-id');
    expect(tree.find('EuiFormLabel').prop('htmlFor')).toEqual('generated-id');

    // Input is described by help and error text.
    expect(tree.find('EuiFormHelpText').prop('id')).toEqual(
      'generated-id-help-0'
    );
    expect(tree.find('EuiFormErrorText').at(0).prop('id')).toEqual(
      'generated-id-error-0'
    );
    expect(tree.find('EuiFormErrorText').at(1).prop('id')).toEqual(
      'generated-id-error-1'
    );
    expect(tree.find('input').prop('aria-describedby')).toEqual(
      'generated-id-help-0 generated-id-error-0 generated-id-error-1'
    );
  });

  describe('props', () => {
    test('label is rendered', () => {
      const component = shallow(
        <EuiFormRow label="label">
          <input />
        </EuiFormRow>
      );

      expect(component).toMatchSnapshot();
    });

    test('label append is rendered', () => {
      const component = shallow(
        <EuiFormRow label="label" labelAppend="append">
          <input />
        </EuiFormRow>
      );

      expect(component).toMatchSnapshot();
    });

    test('label renders as a legend and subsquently a fieldset wrapper', () => {
      const component = shallow(
        <EuiFormRow label="label" labelType="legend">
          <input />
        </EuiFormRow>
      );

      expect(component).toMatchSnapshot();
    });

    test('describedByIds is rendered', () => {
      const component = shallow(
        <EuiFormRow describedByIds={['generated-id-additional']}>
          <input />
        </EuiFormRow>
      );

      expect(component).toMatchSnapshot();
    });

    test('id is rendered', () => {
      const component = render(
        <EuiFormRow id="id">
          <input />
        </EuiFormRow>
      );

      expect(component).toMatchSnapshot();
    });

    test('isInvalid is rendered', () => {
      const component = render(
        <EuiFormRow isInvalid label="label">
          <input />
        </EuiFormRow>
      );

      expect(component).toMatchSnapshot();
    });

    test('error as string is rendered', () => {
      const component = render(
        <EuiFormRow error="Error" isInvalid={true}>
          <input />
        </EuiFormRow>
      );

      expect(component).toMatchSnapshot();
    });

    test('error as array is rendered', () => {
      const component = render(
        <EuiFormRow error={['Error', 'Error2']} isInvalid={true}>
          <input />
        </EuiFormRow>
      );

      expect(component).toMatchSnapshot();
    });

    test('error is not rendered if isInvalid is false', () => {
      const component = render(
        <EuiFormRow error={['Error']} isInvalid={false}>
          <input />
        </EuiFormRow>
      );

      expect(component).toMatchSnapshot();
    });

    test('helpText is rendered', () => {
      const component = render(
        <EuiFormRow helpText={<span>This is help text.</span>}>
          <input />
        </EuiFormRow>
      );

      expect(component).toMatchSnapshot();
    });

    test('hasEmptyLabelSpace is rendered', () => {
      const component = render(
        <EuiFormRow hasEmptyLabelSpace>
          <input />
        </EuiFormRow>
      );

      expect(component).toMatchSnapshot();
    });

    test('fullWidth is rendered', () => {
      const component = render(
        <EuiFormRow fullWidth>
          <input />
        </EuiFormRow>
      );

      expect(component).toMatchSnapshot();
    });

    describe('display type', () => {
      DISPLAYS.forEach((display) => {
        test(`${display} is rendered`, () => {
          const component = render(
            <EuiFormRow display={display}>
              <input />
            </EuiFormRow>
          );

          expect(component).toMatchSnapshot();
        });
      });
    });
  });

  describe('behavior', () => {
    describe('onFocus', () => {
      test('is called in child', () => {
        const focusMock = jest.fn();

        const component = mount(
          <EuiFormRow label={<span>Label</span>}>
            <input onFocus={focusMock} />
          </EuiFormRow>
        );

        component.find('input').simulate('focus');

        expect(focusMock).toBeCalledTimes(1);

        // Ensure the focus event is properly fired on the parent
        // which will pass down to the EuiFormLabel
        expect(component).toMatchSnapshot();
      });

      test('works in parent even if not in child', () => {
        const component = mount(
          <EuiFormRow label={<span>Label</span>}>
            <input />
          </EuiFormRow>
        );

        component.find('input').simulate('focus');

        // Ensure the focus event is properly fired on the parent
        // which will pass down to the EuiFormLabel
        expect(component).toMatchSnapshot();
      });
    });

    describe('onBlur', () => {
      test('is called in child', () => {
        const blurMock = jest.fn();

        const component = mount(
          <EuiFormRow label={<span>Label</span>}>
            <input onBlur={blurMock} />
          </EuiFormRow>
        );

        component.find('input').simulate('blur');

        expect(blurMock).toBeCalledTimes(1);

        // Ensure the blur event is properly fired on the parent
        // which will pass down to the EuiFormLabel
        expect(component).toMatchSnapshot();
      });

      test('works in parent even if not in child', () => {
        const component = mount(
          <EuiFormRow label={<span>Label</span>}>
            <input />
          </EuiFormRow>
        );

        component.find('input').simulate('blur');

        // Ensure the blur event is properly fired on the parent
        // which will pass down to the EuiFormLabel
        expect(component).toMatchSnapshot();
      });
    });
  });
});
