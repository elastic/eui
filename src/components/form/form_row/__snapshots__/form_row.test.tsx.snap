// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EuiFormRow behavior onBlur is called in child 1`] = `
<EuiFormRow
  describedByIds={Array []}
  display="row"
  fullWidth={false}
  hasChildLabel={true}
  hasEmptyLabelSpace={false}
  label={
    <span>
      Label
    </span>
  }
  labelType="label"
>
  <div
    className="euiFormRow"
    id="generated-id-row"
  >
    <div
      className="euiFormRow__labelWrapper"
    >
      <EuiFormLabel
        className="euiFormRow__label"
        htmlFor="generated-id"
        isFocused={false}
        type="label"
      >
        <label
          className="euiFormLabel euiFormRow__label"
          htmlFor="generated-id"
        >
          <span>
            Label
          </span>
        </label>
      </EuiFormLabel>
    </div>
    <div
      className="euiFormRow__fieldWrapper"
    >
      <input
        id="generated-id"
        onBlur={[Function]}
        onFocus={[Function]}
      />
    </div>
  </div>
</EuiFormRow>
`;

exports[`EuiFormRow behavior onBlur works in parent even if not in child 1`] = `
<EuiFormRow
  describedByIds={Array []}
  display="row"
  fullWidth={false}
  hasChildLabel={true}
  hasEmptyLabelSpace={false}
  label={
    <span>
      Label
    </span>
  }
  labelType="label"
>
  <div
    className="euiFormRow"
    id="generated-id-row"
  >
    <div
      className="euiFormRow__labelWrapper"
    >
      <EuiFormLabel
        className="euiFormRow__label"
        htmlFor="generated-id"
        isFocused={false}
        type="label"
      >
        <label
          className="euiFormLabel euiFormRow__label"
          htmlFor="generated-id"
        >
          <span>
            Label
          </span>
        </label>
      </EuiFormLabel>
    </div>
    <div
      className="euiFormRow__fieldWrapper"
    >
      <input
        id="generated-id"
        onBlur={[Function]}
        onFocus={[Function]}
      />
    </div>
  </div>
</EuiFormRow>
`;

exports[`EuiFormRow behavior onFocus is called in child 1`] = `
<EuiFormRow
  describedByIds={Array []}
  display="row"
  fullWidth={false}
  hasChildLabel={true}
  hasEmptyLabelSpace={false}
  label={
    <span>
      Label
    </span>
  }
  labelType="label"
>
  <div
    className="euiFormRow"
    id="generated-id-row"
  >
    <div
      className="euiFormRow__labelWrapper"
    >
      <EuiFormLabel
        className="euiFormRow__label"
        htmlFor="generated-id"
        isFocused={true}
        type="label"
      >
        <label
          className="euiFormLabel euiFormRow__label euiFormLabel-isFocused"
          htmlFor="generated-id"
        >
          <span>
            Label
          </span>
        </label>
      </EuiFormLabel>
    </div>
    <div
      className="euiFormRow__fieldWrapper"
    >
      <input
        id="generated-id"
        onBlur={[Function]}
        onFocus={[Function]}
      />
    </div>
  </div>
</EuiFormRow>
`;

exports[`EuiFormRow behavior onFocus works in parent even if not in child 1`] = `
<EuiFormRow
  describedByIds={Array []}
  display="row"
  fullWidth={false}
  hasChildLabel={true}
  hasEmptyLabelSpace={false}
  label={
    <span>
      Label
    </span>
  }
  labelType="label"
>
  <div
    className="euiFormRow"
    id="generated-id-row"
  >
    <div
      className="euiFormRow__labelWrapper"
    >
      <EuiFormLabel
        className="euiFormRow__label"
        htmlFor="generated-id"
        isFocused={true}
        type="label"
      >
        <label
          className="euiFormLabel euiFormRow__label euiFormLabel-isFocused"
          htmlFor="generated-id"
        >
          <span>
            Label
          </span>
        </label>
      </EuiFormLabel>
    </div>
    <div
      className="euiFormRow__fieldWrapper"
    >
      <input
        id="generated-id"
        onBlur={[Function]}
        onFocus={[Function]}
      />
    </div>
  </div>
</EuiFormRow>
`;

exports[`EuiFormRow is rendered 1`] = `
<div
  aria-label="aria-label"
  class="euiFormRow testClass1 testClass2"
  data-test-subj="test subject string"
  id="generated-id-row"
>
  <div
    class="euiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`EuiFormRow props describedByIds is rendered 1`] = `
<div
  className="euiFormRow"
  id="generated-id-row"
>
  <div
    className="euiFormRow__fieldWrapper"
  >
    <input
      aria-describedby="generated-id-additional"
      id="generated-id"
      onBlur={[Function]}
      onFocus={[Function]}
    />
  </div>
</div>
`;

exports[`EuiFormRow props display type center is rendered 1`] = `
<div
  class="euiFormRow"
  id="generated-id-row"
>
  <div
    class="euiFormRow__fieldWrapper euiFormRow__fieldWrapperDisplayOnly"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`EuiFormRow props display type centerCompressed is rendered 1`] = `
<div
  class="euiFormRow euiFormRow--compressed"
  id="generated-id-row"
>
  <div
    class="euiFormRow__fieldWrapper euiFormRow__fieldWrapperDisplayOnly"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`EuiFormRow props display type columnCompressed is rendered 1`] = `
<div
  class="euiFormRow euiFormRow--compressed euiFormRow--horizontal"
  id="generated-id-row"
>
  <div
    class="euiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`EuiFormRow props display type columnCompressedSwitch is rendered 1`] = `
<div
  class="euiFormRow euiFormRow--compressed euiFormRow--horizontal euiFormRow--hasSwitch"
  id="generated-id-row"
>
  <div
    class="euiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`EuiFormRow props display type row is rendered 1`] = `
<div
  class="euiFormRow"
  id="generated-id-row"
>
  <div
    class="euiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`EuiFormRow props display type rowCompressed is rendered 1`] = `
<div
  class="euiFormRow euiFormRow--compressed"
  id="generated-id-row"
>
  <div
    class="euiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`EuiFormRow props error as array is rendered 1`] = `
<div
  class="euiFormRow"
  id="generated-id-row"
>
  <div
    class="euiFormRow__fieldWrapper"
  >
    <input
      aria-describedby="generated-id-error-0 generated-id-error-1"
      id="generated-id"
    />
    <div
      aria-live="polite"
      class="euiFormErrorText euiFormRow__text"
      id="generated-id-error-0"
    >
      Error
    </div>
    <div
      aria-live="polite"
      class="euiFormErrorText euiFormRow__text"
      id="generated-id-error-1"
    >
      Error2
    </div>
  </div>
</div>
`;

exports[`EuiFormRow props error as string is rendered 1`] = `
<div
  class="euiFormRow"
  id="generated-id-row"
>
  <div
    class="euiFormRow__fieldWrapper"
  >
    <input
      aria-describedby="generated-id-error-0"
      id="generated-id"
    />
    <div
      aria-live="polite"
      class="euiFormErrorText euiFormRow__text"
      id="generated-id-error-0"
    >
      Error
    </div>
  </div>
</div>
`;

exports[`EuiFormRow props error is not rendered if isInvalid is false 1`] = `
<div
  class="euiFormRow"
  id="generated-id-row"
>
  <div
    class="euiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`EuiFormRow props fullWidth is rendered 1`] = `
<div
  class="euiFormRow euiFormRow--fullWidth"
  id="generated-id-row"
>
  <div
    class="euiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`EuiFormRow props hasEmptyLabelSpace is rendered 1`] = `
<div
  class="euiFormRow euiFormRow--hasEmptyLabelSpace"
  id="generated-id-row"
>
  <div
    class="euiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`EuiFormRow props helpText is rendered 1`] = `
<div
  class="euiFormRow"
  id="generated-id-row"
>
  <div
    class="euiFormRow__fieldWrapper"
  >
    <input
      aria-describedby="generated-id-help-0"
      id="generated-id"
    />
    <div
      class="euiFormHelpText euiFormRow__text"
      id="generated-id-help-0"
    >
      <span>
        This is help text.
      </span>
    </div>
  </div>
</div>
`;

exports[`EuiFormRow props id is rendered 1`] = `
<div
  class="euiFormRow"
  id="id-row"
>
  <div
    class="euiFormRow__fieldWrapper"
  >
    <input
      id="id"
    />
  </div>
</div>
`;

exports[`EuiFormRow props isInvalid is rendered 1`] = `
<div
  class="euiFormRow"
  id="generated-id-row"
>
  <div
    class="euiFormRow__labelWrapper"
  >
    <label
      aria-invalid="true"
      class="euiFormLabel euiFormRow__label euiFormLabel-isInvalid"
      for="generated-id"
    >
      label
    </label>
  </div>
  <div
    class="euiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`EuiFormRow props label append is rendered 1`] = `
<div
  className="euiFormRow"
  id="generated-id-row"
>
  <div
    className="euiFormRow__labelWrapper"
  >
    <EuiFormLabel
      className="euiFormRow__label"
      htmlFor="generated-id"
      isFocused={false}
      type="label"
    >
      label
    </EuiFormLabel>
     
    append
  </div>
  <div
    className="euiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
      onBlur={[Function]}
      onFocus={[Function]}
    />
  </div>
</div>
`;

exports[`EuiFormRow props label is rendered 1`] = `
<div
  className="euiFormRow"
  id="generated-id-row"
>
  <div
    className="euiFormRow__labelWrapper"
  >
    <EuiFormLabel
      className="euiFormRow__label"
      htmlFor="generated-id"
      isFocused={false}
      type="label"
    >
      label
    </EuiFormLabel>
  </div>
  <div
    className="euiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
      onBlur={[Function]}
      onFocus={[Function]}
    />
  </div>
</div>
`;

exports[`EuiFormRow props label renders as a legend and subsquently a fieldset wrapper 1`] = `
<fieldset
  className="euiFormRow"
  id="generated-id-row"
>
  <div
    className="euiFormRow__labelWrapper"
  >
    <EuiFormLabel
      className="euiFormRow__label"
      type="legend"
    >
      label
    </EuiFormLabel>
  </div>
  <div
    className="euiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
      onBlur={[Function]}
      onFocus={[Function]}
    />
  </div>
</fieldset>
`;
